File: bindings/python/algorithms/invariants.cpp
#include <boost/python.hpp>

#include "../../src/chemist/molecule.h"
#include "../../src/algorithms/invariants.h"
#include "../common.h"

using Helium::Chemist::Molecule;
using namespace boost::python;

struct AtomInvariantWrapper : AtomInvariant, wrapper<AtomInvariant>
{
  unsigned long call_1(const Molecule &mol, const Molecule::atom_type &atom) const
  {
    return this->get_override("__call__")(boost::ref(mol), boost::ref(atom));
  }
};

struct BondInvariantWrapper : BondInvariant, wrapper<BondInvariant>
{
  unsigned long call_1(const Molecule &mol, const Molecule::bond_type &bond) const
  {
    return this->get_override("__call__")(boost::ref(mol), boost::ref(bond));
  }
};

unsigned long DefaultAtomInvariant_call(const Helium::DefaultAtomInvariant &invariant, const Molecule &mol,
    const Molecule::atom_type &atom)
{
  return invariant(mol, atom);
}

unsigned long DefaultBondInvariant_call(const Helium::DefaultBondInvariant &invariant, const Molecule &mol,
    const Molecule::bond_type &bond)
{
  return invariant(mol, bond);
}

void export_default_atom_invariant()
{
  scope in_DefaultAtomInvariant = class_<Helium::DefaultAtomInvariant, boost::noncopyable>("DefaultAtomInvariant", init<>())
    .def(init<int>())
    .def("__call__", &DefaultAtomInvariant_call)
    ;

  enum_<Helium::DefaultAtomInvariant::Invariants>("Invariants")
    .value("None", Helium::DefaultAtomInvariant::None)
    .value("Element", Helium::DefaultAtomInvariant::Element)
    .value("Mass", Helium::DefaultAtomInvariant::Mass)
    .value("Charge", Helium::DefaultAtomInvariant::Charge)
    .value("Degree", Helium::DefaultAtomInvariant::Degree)
    .value("Aromatic", Helium::DefaultAtomInvariant::Aromatic)
    .value("All", Helium::DefaultAtomInvariant::All)
    ;
}

void export_default_bond_invariant()
{
  scope in_DefaultBondInvariant = class_<Helium::DefaultBondInvariant, boost::noncopyable>("DefaultBondInvariant", init<>())
    .def(init<int>())
    .def("__call__", &DefaultBondInvariant_call)
    ;

  enum_<Helium::DefaultBondInvariant::Invariants>("Invariants")
    .value("None", Helium::DefaultBondInvariant::None)
    .value("Order", Helium::DefaultBondInvariant::Order)
    .value("Aromatic", Helium::DefaultBondInvariant::Aromatic)
    .value("All", Helium::DefaultBondInvariant::All)
    ;
}

void export_invariants()
{

  class_<AtomInvariantWrapper, boost::noncopyable>("AtomInvariant")
    .def("__call__", &AtomInvariant::call_1);
  class_<BondInvariantWrapper, boost::noncopyable>("BondInvariant")
    .def("__call__", &BondInvariant::call_1);

  export_default_atom_invariant();
  export_default_bond_invariant();

}

File: test/python/invariants.py
import helium
import unittest

SMILES = helium.Smiles()

class TestInvariants(unittest.TestCase):

    def test_default_atom_invariant(self):
        mol = helium.Molecule()
        SMILES.read('CNO', mol)

        invariant = helium.DefaultAtomInvariant()
        self.assertEqual(2110012006, invariant(mol, mol.atom(0)))

        invariant = helium.DefaultAtomInvariant(helium.DefaultAtomInvariant.Invariants.Element)
        self.assertEqual(6, invariant(mol, mol.atom(0)))

        invariant = helium.DefaultAtomInvariant(helium.DefaultAtomInvariant.Invariants.Mass)
        self.assertEqual(12000, invariant(mol, mol.atom(0)))

        ec = helium.extended_connectivities(mol, invariant)
        self.assertTrue(isinstance(ec, list))
        self.assertEqual(3, len(ec))

    def test_default_bond_invariant(self):
        mol = helium.Molecule()
        SMILES.read('CN=O:C', mol)

        invariant = helium.DefaultBondInvariant()
        self.assertEqual(1, invariant(mol, mol.bond(0)))

        invariant = helium.DefaultBondInvariant(helium.DefaultBondInvariant.Invariants.Order)
        self.assertEqual(2, invariant(mol, mol.bond(1)))

        invariant = helium.DefaultBondInvariant(helium.DefaultBondInvariant.Invariants.Aromatic)
        self.assertEqual(100, invariant(mol, mol.bond(2)))


if __name__ == '__main__':
    unittest.main()
